### ISSUE TITLE
Migration error with Mysql 5.7: "Cannot add foreign key constraint"

### ISSUE THREAD

**Issue by Toub** [oTree-org/otree-core#454]
I just made otree to work with MySQL, but I had some errors during the schema migration.

My install is described here: https://github.com/oTree-org/otree-docs/issues/20

I tried both with otree migrate or python manage.py migrate, is there any difference?

```
otree migrate
Operations to perform:
  Synchronize unmigrated apps: bargaining, beauty, ultimatum, prisoner, lemon_market, messages, public_goods_simple, matching_pennies_tutorial, stag_hunt, staticfiles, floppyforms, trust_simple, traveler_dilemma, timeout, real_effort, vickrey_auction, battle_of_the_sexes, models_concrete, payment_info, matching_pennies, sslserver, corsheaders, stackelberg_competition, dictator, common_value_auction, rest_framework, idmap, public_goods, bertrand_competition, volunteer_dilemma, principal_agent, cournot_competition, trust, survey
  Apply all migrations: otree, kombu_transport_django, admin, djcelery, sessions, auth, contenttypes
Synchronizing apps without migrations:
  Creating tables...
    Creating table corsheaders_corsmodel
    Creating table bargaining_subsession
    Creating table bargaining_group
    Creating table bargaining_player
    Creating table beauty_subsession
    Creating table beauty_group
    Creating table beauty_player
    Creating table ultimatum_subsession
    Creating table ultimatum_group
    Creating table ultimatum_player
    Creating table prisoner_subsession
    Creating table prisoner_group
    Creating table prisoner_player
    Creating table lemon_market_subsession
    Creating table lemon_market_group
    Creating table lemon_market_player
    Creating table public_goods_simple_subsession
    Creating table public_goods_simple_group
    Creating table public_goods_simple_player
    Creating table matching_pennies_tutorial_subsession
    Creating table matching_pennies_tutorial_group
    Creating table matching_pennies_tutorial_player
    Creating table stag_hunt_subsession
    Creating table stag_hunt_group
    Creating table stag_hunt_player
    Creating table trust_simple_subsession
    Creating table trust_simple_group
    Creating table trust_simple_player
    Creating table traveler_dilemma_subsession
    Creating table traveler_dilemma_group
    Creating table traveler_dilemma_player
    Creating table real_effort_subsession
    Creating table real_effort_group
    Creating table real_effort_player
    Creating table vickrey_auction_subsession
    Creating table vickrey_auction_group
    Creating table vickrey_auction_player
    Creating table battle_of_the_sexes_subsession
    Creating table battle_of_the_sexes_group
    Creating table battle_of_the_sexes_player
    Creating table payment_info_subsession
    Creating table payment_info_group
    Creating table payment_info_player
    Creating table matching_pennies_subsession
    Creating table matching_pennies_group
    Creating table matching_pennies_player
    Creating table stackelberg_competition_subsession
    Creating table stackelberg_competition_group
    Creating table stackelberg_competition_player
    Creating table dictator_subsession
    Creating table dictator_group
    Creating table dictator_player
    Creating table common_value_auction_subsession
    Creating table common_value_auction_group
    Creating table common_value_auction_player
    Creating table public_goods_subsession
    Creating table public_goods_group
    Creating table public_goods_player
    Creating table bertrand_competition_subsession
    Creating table bertrand_competition_group
    Creating table bertrand_competition_player
    Creating table volunteer_dilemma_subsession
    Creating table volunteer_dilemma_group
    Creating table volunteer_dilemma_player
    Creating table principal_agent_subsession
    Creating table principal_agent_group
    Creating table principal_agent_player
    Creating table cournot_competition_subsession
    Creating table cournot_competition_group
    Creating table cournot_competition_player
    Creating table trust_subsession
    Creating table trust_group
    Creating table trust_player
    Creating table survey_subsession
    Creating table survey_group
    Creating table survey_player
    Running deferred SQL...
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/django/base.py", line 177, in _execute_wrapper
    return method(query, args)
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/cursor.py", line 515, in execute
    self._handle_result(self._connection.cmd_query(stmt))
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/connection.py", line 488, in cmd_query
    result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/connection.py", line 395, in _handle_result
    raise errors.get_exception(packet)
mysql.connector.errors.DatabaseError: 1215 (HY000): Cannot add foreign key constraint

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/otree", line 9, in <module>
    load_entry_point('otree-core==0.4.43', 'console_scripts', 'otree')()
  File "/app/sources/core/otree/management/cli.py", line 220, in otree_cli
    execute_from_command_line(argv, 'otree')
  File "/app/sources/core/otree/management/cli.py", line 176, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.5/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/migrate.py", line 179, in handle
    created_models = self.sync_apps(connection, executor.loader.unmigrated_apps)
  File "/usr/local/lib/python3.5/site-packages/django/core/management/commands/migrate.py", line 318, in sync_apps
    cursor.execute(statement)
  File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 62, in execute
    return self.cursor.execute(sql)
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/django/base.py", line 227, in execute
    return self._execute_wrapper(self.cursor.execute, query, new_args)
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/django/base.py", line 195, in _execute_wrapper
    utils.DatabaseError(err.msg), sys.exc_info()[2])
  File "/usr/local/lib/python3.5/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/django/base.py", line 177, in _execute_wrapper
    return method(query, args)
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/cursor.py", line 515, in execute
    self._handle_result(self._connection.cmd_query(stmt))
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/connection.py", line 488, in cmd_query
    result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))
  File "/usr/local/lib/python3.5/site-packages/mysql/connector/connection.py", line 395, in _handle_result
    raise errors.get_exception(packet)
django.db.utils.DatabaseError: Cannot add foreign key constraint
```

Then, if I run python manage.py migrate again, the process ends successfully:

```
$ python manage.py migrate
Operations to perform:
  Synchronize unmigrated apps: trust_simple, principal_agent, messages, vickrey_auction, beauty, common_value_auction, traveler_dilemma, prisoner, corsheaders, timeout, bertrand_competition, public_goods, models_concrete, trust, idmap, matching_pennies_tutorial, real_effort, battle_of_the_sexes, public_goods_simple, dictator, sslserver, volunteer_dilemma, floppyforms, ultimatum, stag_hunt, staticfiles, lemon_market, rest_framework, cournot_competition, stackelberg_competition, bargaining, payment_info, survey, matching_pennies
  Apply all migrations: admin, djcelery, contenttypes, auth, kombu_transport_django, otree, sessions
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
  Installing custom SQL...
Running migrations:
  Rendering model states... DONE
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying djcelery.0001_initial... OK
  Applying kombu_transport_django.0001_initial... OK
  Applying otree.0001_initial... OK
  Applying sessions.0001_initial... OK
[INFO|2016-04-12 12:26:41,783] otree > Creating default superuser. Username: admin

```

Looks like the interesting line is here: 

``` python
django.db.utils.DatabaseError: Cannot add foreign key constraint
```

I don't know if only one or several indexes are missing, but the application is running and I can create new sessions stored in mysql.

Any help about this problem will be appreciated.

**Comment by oTree-org** [comment]
The problem might be that your apps in your project folder don't have migrations, but they depend on otree-core, which has migrations. This can break:

https://docs.djangoproject.com/en/1.8/topics/migrations/#dependencies

Try running "otree makemigrations foo" for each app foo in your project (e.g. bargaining, beauty, ultimatum, prisoner, etc...). They are listed as 'unmigrated apps" in the output you posted above.

Then delete the database and run "otree migrate".

**Comment by Toub** [comment]
Thanks for your answer.

I understand better how migrations works now, but for some reason, the dependency order for migrations does not work in my case.

I limited the SESSION_CONFIGS to only one app "public_goods", then created the migration files, who contains dependency to "otree", but the error still occurs.

I tried several times and the list of applications to migrate in the list looks like "random".

However, the following workaround works:
* drop/create database
* remove all applications from SESSION_CONFIGS in my settings.py
* run "otree migrate" to migrate otree core
* add one or all applications to SESSION_CONFIGS in my settings.py
* run "otree migrate" to migrate the installed otree games

Any idea?
